cognite/client/_api/datapoints.py

```diff
@@ -1,117 +1,901 @@
+            if res is None:
+                continue
+            # We may dynamically split subtasks based on what % of time range was returned:
+            if new_subtasks := subtask.store_partial_result(res):
+                self._queue_new_subtasks(pool, futures_dct, new_subtasks)
+            if ts_task.is_done:  # "Parent" ts task might be done before a subtask is finished
+                if all(parent.is_done for parent in ts_task_lookup.values()):
+                    pool.shutdown(wait=False)
+                    break
+                if ts_task.has_limit:
+                    # For finished limited queries, cancel all unstarted futures for same parent:
+                    self._cancel_futures_for_finished_ts_task(ts_task, futures_dct)
+                continue
+            elif subtask.is_done:
+                continue
+            self._queue_new_subtasks(pool, futures_dct, [subtask])
+        # Return only non-missing time series tasks in correct order given by `all_queries`:
+        return list(filter(None, map(ts_task_lookup.get, self.all_queries)))
+
+    def _create_initial_tasks(
+        self,
+        pool: PriorityThreadPoolExecutor,
+        use_numpy: bool,
+    ) -> Tuple[Dict[Future, BaseDpsFetchSubtask], Dict[_SingleTSQueryBase, BaseConcurrentTask]]:
+        futures_dct: Dict[Future, BaseDpsFetchSubtask] = {}
+        ts_task_lookup, payload = {}, {"ignoreUnknownIds": False}
+        for query in self.all_queries:
+            ts_task = ts_task_lookup[query] = query.ts_task_type(query=query, eager_mode=True, use_numpy=use_numpy)
+            for subtask in ts_task.split_into_subtasks(self.max_workers, self.n_queries):
+                future = pool.submit(self.__request_datapoints_jit, subtask, payload, priority=subtask.priority)
+                futures_dct[future] = subtask
+        return futures_dct, ts_task_lookup
+
+    def _queue_new_subtasks(
+        self,
+        pool: PriorityThreadPoolExecutor,
+        futures_dct: Dict[Future, BaseDpsFetchSubtask],
+        new_subtasks: Sequence[BaseDpsFetchSubtask],
+    ) -> None:
+        for task in new_subtasks:
+            future = pool.submit(self.__request_datapoints_jit, task, priority=task.priority)
+            futures_dct[future] = task
+
+    def _get_result_with_exception_handling(
+        self,
+        future: Future,
+        ts_task: BaseConcurrentTask,
+        ts_task_lookup: Dict[_SingleTSQueryBase, BaseConcurrentTask],
+        futures_dct: Dict[Future, BaseDpsFetchSubtask],
+    ) -> Optional[DataPointListItem]:
+        try:
+            if (res := future.result()) is not None:
+                return res.items[0]
+            return None
+        except CancelledError:
+            return None
+        except CogniteAPIError as e:
+            if not (e.code == 400 and e.missing and ts_task.query.ignore_unknown_ids):
+                collect_exc_info_and_raise([e])
```

Comment:
I just realized this might be confusing to users; at least it’s inconsistent. In its current form, when asking for “few time series” (eager fetching), only the first ts that doesn’t exist will raise, as opposed to “many ts” (chunked fetching) which will raise a combined exception listing all missing time series (not allowed to not exist). Thoughts @erlendvollset ?

Reply:
Let's not address this now. We will rework exception handling at some point and then consistently use more well-established patterns for exceptions in concurrent scenarios (i.e. task/exception groups).

Reply:
Sounds good. A simple alternative is to append some text to `e.message`, but it might be confusing so I'll just leave it  be.